micro_http/src/server.rs

```diff
@@ -0,0 +1,551 @@
+/// there is a notification on that fd, `wait_server` should be called
+/// once.
+pub struct HttpServer {
+    /// Socket on which we listen for new connections.
+    socket: UnixListener,
+    /// File descriptor of the server's epoll structure.
+    epoll_fd: RawFd,
+    /// Holds the token-connection pairs of the server.
+    /// Each connection has an associated token.
+    connections: HashMap<Token, ConnectionWithFd<UnixStream>>,
+    /// Holds all available tokens which are not currently
+    /// associated with a connection.
+    token_queue: VecDeque<Token>,
+}
+
+impl HttpServer {
+    /// Constructor for `HttpServer`.
+    pub fn new<P: AsRef<Path>>(path_to_socket: P) -> Result<Self> {
+        let socket = UnixListener::bind(path_to_socket).map_err(|_| ServerError::IOError)?;
+        let epoll_fd = epoll::create(false).map_err(|_| ServerError::IOError)?;
+        Ok(HttpServer {
+            socket,
+            epoll_fd,
+            connections: HashMap::new(),
+            token_queue: (0..MAX_CONNECTIONS as u64).collect(),
+        })
+    }
+
+    /// Starts the HTTP Server.
+    pub fn start_server(&mut self) -> Result<()> {
+        epoll::ctl(
+            self.epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.socket.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, NEW_CONNECTION_EPOLL_TOKEN),
+        )
+        .map_err(|_| ServerError::IOError)?;
+
+        Ok(())
+    }
+
+    /// Calls `epoll::wait` on the epoll structure containing all open connections.
+    ///
+    /// This function is responsible for the data exchange with the clients and should
+    /// be called when we are either notified through `epoll` that we need to exchange
+    /// data with at least a client or when we don't need to perform any other operations
+    /// on this thread and we can afford to call it in a loop.
+    ///
+    /// Returns a collection of complete and valid requests to be processed by the user
+    /// of the server. Once processed, responses should be sent using `enqueue_responses`.
+    pub fn wait_server(&mut self) -> Result<Vec<(Token, Request)>> {
+        let mut parsed_requests: Vec<(Token, Request)> = vec![];
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); MAX_CONNECTIONS];
+        let num_ev =
+            epoll::wait(self.epoll_fd, -1, &mut events[..]).map_err(|_| ServerError::IOError)?;
+        let mut streams_to_be_removed: Vec<Token> = vec![];
+        for e in events.iter().take(num_ev) {
+            match e.data as Token {
+                NEW_CONNECTION_EPOLL_TOKEN => {
+                    match self.handle_new_connection() {
+                        Err(ServerError::ServerFull) => continue,
+                        Err(error) => return Err(error),
+                        Ok(()) => {}
+                    };
+                }
+                token => {
+                    self.serve_connection(
```

Comment:
Instead of the `self.serve_connection(token, e.events, &mut streams_to_be_removed, &mut parsed_requests)?;`, get the `conn` from the map then do `parsed_requests.append(conn.process(e.events));` here.